src/github.com/stellar/horizon/resource/account.go

```diff
@@ -71,3 +72,16 @@ func (this *Account) Populate(
 	this.Links.Data.PopulateTemplated()
 	return
 }
+
+// GetData returns decoded value for a given key. If the key does
+// not exist, empty slice will be returned.
+func (this *Account) GetData(key string) []byte {
+	data, exists := this.Data[key]
+	if !exists {
+		return nil
+	}
+
+	// We assume that horizon always returns valid base64 encoded string
+	decoded, _ := base64.StdEncoding.DecodeString(data)
```

Comment:
> If many of the things I listed are validated by horizon, then why can't we validate data value too (valid base64) and allow developers to decode it with assumption that it's correct? (so GetData without returning error). I understand that returning an error is another layer of defence but then:

We can't validate the data value ahead of time because it is being loaded from stellar-core's database.  It is an external system to horizon and can contain anything as far as horizon is concerned.  stellar-core could introduce a bug that corrupts that field; we have no way of ensuring that data is and will always be valid.  It isn't the same as the history tables, which are logically owned by horizon since horizon is populating them.  I don't know how more plainly to put this and honestly am baffled that you don't think we should be validating stellar-core data wherever we can.  The fewer instances of unconditional trust in an external system we have, the better.

> If we should check values returned by horizon in the client code, then why don't we already check if account ids, hashes, links and other things I listed are valid? We should also advice developers to check the full history of sender account when a payment arrives because horizon can send incorrect value because of a bug.


You are treating this like its one or the other. Either validate everything or validate nothing.  That's not the case:  whether we choose to validate something when it is loaded from outside the process should be answered on a case by case basis.  Here's some of the questions you should be asking yourself:

- Is it feasible to validate the value?  For example, we have no choice but to trust the balance fields from the stellar-core database.  There's no easy way that I know of the validate a balance other than contacting an independent stellar-core instance. 
- Can the value be validated further downstream?  If so, perhaps you should only validate it if you are operating upon it yourself.  In this case, you are stripping away the base64 so the downstream can no longer validate the value.
- Is the value produced by your application? If so, then you can often get good enough assurance that a value will be value by using your test suite and a set of validations that run when you write data to the database.  ActiveRecord gets by pretty well using this model, but even still you see database administrators arguing for check constraints on columns to enforce the data model.
- If the value is external, was it provided to you unasked (via a client http request, for example)?  If so, IMO, you _must_ validate it in some way before further operating on the value.  SQL injection lives here.

There are more I can go into, but I think I've made my point.

Here's the bottom line, IMO:

If you cannot prove a value's validity, then the core question you should be asking concerns your apetite for risk.  Are you comfortable with the possible outcomes should the value in question value be corrupted?

In my opinion, we should trust but verify stellar-core's work where we can.  When I import data, I ensure the ledger chain is unbroken.  When I decode XDR data, I ensure we consume all of the data so that our confidence that the decoded value is genuine is raised.  The coverage isn't complete for sure, but the goal remains.

---

Let me respond directly to you questions:

> then why don't we already check if account ids, hashes, links and other things I listed are valid?

Why don't we validate account ids?  Because they're opaque values most of the time that we just pass on to the next consumer.  The horizon client can validate them if they need to.  Just as when we need to, we do.

Hashes?  Same thing.

Links?  We do validate them:  https://github.com/stellar/horizon/blob/master/src/github.com/stellar/horizon/render/hal/link_builder.go#L46

Posted transactions?  We do validate them: https://github.com/stellar/horizon/blob/master/src/github.com/stellar/horizon/txsub/system.go#L93

---

I wish I had an easy answer to give you, but there simply isn't one, IMO.  We are building sensitive software and we should stive to make invalid data impossible, barring that we should ensure that the risk posed is minimal or can be mitigated downstream.